### ISSUE TITLE
type inference doesn't work in async fn that return Box<dyn SomeTrait>

### ISSUE THREAD

**Issue by macpp** [rust-lang/rust#60424]
i have following fn that works completly fine:
```rust
pub fn test1() -> Box<dyn Debug> {
    Box::new("asdf")
}
```
if i try trivialy make it async just by adding `async` keyword i get build error:
```
error[E0271]: type mismatch resolving `<impl core::future::future::Future as core::future::future::Future>::Output == std::boxed::Box<(dyn std::fmt::Debug + 'static)>`
  --> comix_downloader_lib_async/src/lib.rs:49:25
   |
49 | pub async fn test1() -> Box<dyn Debug> {
   |                         ^^^^^^^^^^^^^^ expected &str, found trait std::fmt::Debug
   |
   = note: expected type `std::boxed::Box<&str>`
              found type `std::boxed::Box<(dyn std::fmt::Debug + 'static)>`
   = note: the return type of a function must have a statically known size
```
i can work around this by defining types of variables explicitly:
```rust
pub async fn test3() -> Box<dyn Debug> {
    let tmp : Box<dyn Debug> = Box::new("asdf");
    tmp
}
``` 
but for obvious reasons this is not as nice as sync version
maybe related to #60414 ?

`rustc --version` : `rustc 1.36.0-nightly (00859e3e6 2019-04-29)`

**Comment by cramertj** [comment]
I'd put this under a similar camp as https://github.com/rust-lang/rust/issues/54326-- the issue here is that we need to fix the output of the `Future` to the return type as provided, rather than trying to infer it from the context (which isn't smart enough in this case to see that a trait object coercion is required).

**Comment by cramertj** [comment]
Related to https://github.com/rust-lang/rust/issues/54326, some discussion in https://rust-lang.zulipchat.com/#narrow/stream/187312-wg-async-foundations/topic/mismatched.20types.20diag.20.2354326/near/168076272.

**Comment by nikomatsakis** [comment]
[Discussed in meeting today](https://rust-lang.zulipchat.com/#narrow/stream/187312-wg-async-foundations/topic/meeting.202019.2E07.2E09/near/170471274). Adding the "unclear" tag: this one is tricky to fix and it's not clear that it should block stabilization. It would definitely be good to have, though. In particular, it seems unlikely to be a forwards compatibility issue, since this is basically a problem of failing to do the coercion that is needed and hence getting more compilation errors (but it would be helpful if people have counterexamples in mind where behavior might change, seems possible).

**Comment by nikomatsakis** [comment]
Discussed in the rust-lang/lang meeting -- we decided this is, ultimately, not a *blocker* but we ought to "try real hard" to fix it.

**Comment by leo60228** [comment]
This even more minimal example has the same problem:
```rust
async fn func(s: &str) -> &[&str] {
    &[s]
}
```

**Comment by cramertj** [comment]
@leo60228 Yup, you need

```rust
async fn func(s: &str) -> &[&str] {
    &[s] as &[&str]
}
```
which correctly generates an error:
```
error[E0515]: cannot return value referencing temporary value
 --> src/lib.rs:2:5
  |
2 |     &[s] as &[&str]
  |     ^---^^^^^^^^^^^
  |     ||
  |     |temporary value created here
  |     returns a value referencing data owned by the current function
```

**Comment by leo60228** [comment]
@cramertj Whoops, I didn't test the non-async version... This does have the problem I intended to demonstrate, though:

```rust
async fn func() -> &'static [&'static str] {
    &["hi"]
}
```

**Comment by blitzerr** [comment]
@nikomatsakis Any notes on this one ?

**Comment by nikomatsakis** [comment]
OK, so I have been doing a bit of digging into this. Our previous efforts to fix those were focused on keeping the existing desugaring. But @Centril was suggesting that we should modify the HIR and I think they are correct. 

Let's walk through what happens here. First off, I'll run with this example:

```rust
async fn func() -> &'static [&'static str] {
    &["hi"]
}
```

This presently gets desugared to an async block, but in the compiler that async block is itself desugared into a call to `from_generator` with a "generator closure" as argument:

```rust
fn func() -> &'static [&'static str] {
  ::std::future::from_generator(gen || {
    &["hi"]
  })
}
```

I think what we want to do, in short, is to generate a return type annotation on that generator. You can see that this will fix the compile error in this [manually desugared version from the playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ade52a1706eadcf5ea620bb1eedd0b85). 

However, I think the right way to do this is *probably* to extend the HIR type with some kind of "current function return type" node for internal use that expands to the return type from the current function. The reason to do this is that we can't just "copy-n-paste" the return type from the function declaration into this position -- the references to lifetimes will not work out correctly, unfortunately. In particular, elided lifetimes are handled by a later pass and they would be treated differently in this position. 

A few more tips. This desugaring is done here:

https://github.com/rust-lang/rust/blob/c7bc0bf82faf9718fa1e59a38f5aab308299ba7d/src/librustc/hir/lowering/item.rs#L1223-L1258

This is a call to [`make_async_expr`]. Note the 3rd argument, `ret_ty`, in particular:

https://github.com/rust-lang/rust/blob/c7bc0bf82faf9718fa1e59a38f5aab308299ba7d/src/librustc/hir/lowering/expr.rs#L451

[`make_async_expr`]: https://github.com/rust-lang/rust/blob/c7bc0bf82faf9718fa1e59a38f5aab308299ba7d/src/librustc/hir/lowering/expr.rs#L447-L454

This argument supplies the return type annotation on the closure. When we invoke it from `async fn`, it is None. We can't just supply `Some`, however, because this expects an *AST* type and we would want to give it a *HIR* type, I think, at least if we're going to be adding some special HIR node representing the "return type of the current function". I have to dig a bit into this part, another option might be to try and convince the lifetime elision code to do the right thing so that we can just copy-and-paste the type.

**Comment by nikomatsakis** [comment]
Ah, actually, I realize something even *easier*. We already supply a flag to the generator code that indicates this is a generator. If we supply a flag that indicates that this is the "async move from an async fn body", we could supply the correct typing hints in type-check.

**Comment by nikomatsakis** [comment]
I may take a stab at this in a few, else I'll poke somebody to give it a try in Zulip.

**Comment by nikomatsakis** [comment]
Oh dang it, I just realized the obvious complication -- the return type of the function is actually a future, so what we really need to do is to extract the `Output` of that. This is going to be an annoying amount of code, though it remains plausible.

**Comment by cramertj** [comment]
I'm sure this has already been considered, but this bug would be trivially fixable if it were possible to have types from the AST appear twice in the HIR. It also seems plausible that such a thing would be useful elsewhere. Is there a sensible path to allowing this?

**Comment by nikomatsakis** [comment]
Got a pending fix in #64999
